File: test/test_search_nocorrect.py
#!/usr/bin/env python3
"""
Test script to investigate the impact of 'nocorrect' parameter on search results.
Tests:
1. Difference between nocorrect=True/False
2. Impact of quoted queries
3. search_suggest method
4. Misspelling detection and correction

Usage:
    python3 test/test_search_nocorrect.py

Required:
    - yandex_music library installed
    - test/token.txt with valid OAuth token
"""

import json
import sys
from pathlib import Path

try:
    from yandex_music import Client
except ImportError:
    print("ERROR: yandex_music library not found. Install it with: pip install yandex-music")
    sys.exit(1)

# Get token from test/token.txt
token_file = Path(__file__).parent / 'token.txt'
if not token_file.exists():
    print(f"ERROR: Token file not found at {token_file}")
    print("Please create test/token.txt with your OAuth token")
    sys.exit(1)

token = token_file.read_text().strip()
if not token:
    print("ERROR: Token file is empty")
    sys.exit(1)

print(f"Token loaded: {token[:20]}...\n")

# Initialize client
try:
    client = Client(token).init()
    print("✓ Client initialized successfully\n")
except Exception as e:
    print(f"ERROR: Failed to initialize client: {e}")
    sys.exit(1)

# Test queries - both correct and misspelled
test_queries = [
    {
        'name': 'Correct spelling',
        'query': 'Гражданская оборона',
        'expected_type': 'artist',
    },
    {
        'name': 'Misspelled (missing letter)',
        'query': 'Гражданская абарона',
        'expected_type': 'artist',
    },
    {
        'name': 'Misspelled (extra letter)',
        'query': 'Би-2 Сонг',
        'expected_type': 'artist',
    },
    {
        'name': 'Quoted query',
        'query': '"Гражданская оборона"',
        'expected_type': 'artist',
    },
]

def print_separator(char='=', length=100):
    """Print a separator line."""
    print(char * length)

def format_number(num):
    """Format number with thousands separator."""
    return f"{num:,}" if num is not None else "N/A"

def search_and_compare(query, search_type='all'):
    """
    Search with nocorrect=False and nocorrect=True and compare results.

    Returns:
        dict: Comparison results
    """
    print(f"\nQuery: '{query}'\n")

    results = {}

    for nocorrect_val in [False, True]:
        try:
            search_result = client.search(
                text=query,
                nocorrect=nocorrect_val,
                type_=search_type,
            )

            if search_result is None:
                print(f"  nocorrect={nocorrect_val}: No results")
                results[nocorrect_val] = None
                continue

            results[nocorrect_val] = {
                'search_result': search_result,
                'counts': {},
            }

            # Count results by type
            for result_type in ['tracks', 'artists', 'albums', 'playlists', 'videos', 'podcasts', 'podcast_episodes', 'users']:
                attr = getattr(search_result, result_type, None)
                count = attr.total if attr else 0
                results[nocorrect_val]['counts'][result_type] = count

            # Check for misspelling
            results[nocorrect_val]['misspell_result'] = search_result.misspell_result
            results[nocorrect_val]['misspell_original'] = search_result.misspell_original
            results[nocorrect_val]['misspell_corrected'] = search_result.misspell_corrected
            results[nocorrect_val]['nocorrect_param'] = search_result.nocorrect

        except Exception as e:
            print(f"  nocorrect={nocorrect_val}: ERROR - {e}")
            results[nocorrect_val] = None

    return results

def print_search_results(query, results):
    """Print formatted search results comparison."""
    print_separator('─', 100)

    # Header
    print(f"{'Category':<25} {'nocorrect=False':<30} {'nocorrect=True':<30}")
    print_separator('─', 100)

    if results[False] and results[True]:
        # Total counts
        for result_type in ['tracks', 'artists', 'albums', 'playlists', 'videos', 'podcasts', 'podcast_episodes', 'users']:
            count_false = results[False]['counts'].get(result_type, 0)
            count_true = results[True]['counts'].get(result_type, 0)

            if count_false > 0 or count_true > 0:
                diff = count_true - count_false
                diff_marker = f" (Δ{diff:+d})" if diff != 0 else ""
                print(f"{result_type:<25} {format_number(count_false):<30} {format_number(count_true):<30}{diff_marker}")

        print_separator('─', 100)

        # Misspelling info
        print(f"{'MISSPELLING INFO':<25}")
        print_separator('─', 100)

        for nocorrect_val in [False, True]:
            res = results[nocorrect_val]
            print(f"\nnocorrect={nocorrect_val}:")
            print(f"  Original query:  {res['misspell_original'] or 'N/A'}")
            print(f"  Corrected to:    {res['misspell_result'] or 'N/A'}")
            print(f"  Was corrected:   {res['misspell_corrected']}")
            print(f"  nocorrect param: {res['nocorrect_param']}")
    else:
        print("ERROR: Could not get results for comparison")

    print("\n")

def test_search_suggest(query_part):
    """Test search_suggest method."""
    print(f"\nTesting search_suggest with query part: '{query_part}'\n")
    print_separator('─', 100)

    try:
        suggestions = client.search_suggest(part=query_part)

        if suggestions is None or not suggestions.suggests:
            print("No suggestions found")
            return

        print(f"Found {len(suggestions.suggests)} suggestions:\n")

        # Group suggestions by type
        by_type = {}
        for i, suggest in enumerate(suggestions.suggests[:10], 1):
            if suggest not in by_type:
                by_type[suggest] = 0
            by_type[suggest] += 1
            print(f"  {i}. {suggest}")

    except Exception as e:
        print(f"ERROR: {e}")

    print("\n")

def test_quoted_queries(queries):
    """Test impact of quotes on search results."""
    print("\n" + "="*100)
    print("TESTING IMPACT OF QUOTES ON SEARCH RESULTS")
    print("="*100 + "\n")

    for query in queries[:2]:  # Test first 2 queries
        unquoted = query
        quoted = f'"{query}"'

        print(f"Comparing quoted vs unquoted: {query}\n")
        print_separator('─', 100)

        results_unquoted = search_and_compare(unquoted, 'track')
        results_quoted = search_and_compare(quoted, 'track')

        print(f"{'Parameter':<30} {'Unquoted':<30} {'Quoted':<30}")
        print_separator('─', 100)

        # Compare basic counts
        if results_unquoted[False] and results_quoted[False]:
            for result_type in ['tracks', 'artists', 'albums']:
                count_unquoted = results_unquoted[False]['counts'].get(result_type, 0)
                count_quoted = results_quoted[False]['counts'].get(result_type, 0)
                diff = count_quoted - count_unquoted
                diff_marker = f" (Δ{diff:+d})" if diff != 0 else ""
                print(f"{result_type:<30} {format_number(count_unquoted):<30} {format_number(count_quoted):<30}{diff_marker}")

        print("\n")

def main():
    """Main test function."""
    print("\n" + "="*100)
    print("YANDEX MUSIC SEARCH NOCORRECT PARAMETER TEST")
    print("="*100)

    # Test 1: Compare nocorrect=True/False
    print("\n" + "="*100)
    print("TEST 1: NOCORRECT PARAMETER IMPACT")
    print("="*100)

    for test_query in test_queries:
        print(f"\n{test_query['name']}:")
        results = search_and_compare(test_query['query'])
        print_search_results(test_query['query'], results)

    # Test 2: search_suggest
    print("\n" + "="*100)
    print("TEST 2: SEARCH_SUGGEST METHOD")
    print("="*100)

    test_search_suggest('граж')
    test_search_suggest('би')

    # Test 3: Impact of quotes
    test_quoted_queries([
        'Гражданская оборона',
        'Bi-2',
        'Pink Floyd',
    ])

    # Test 4: Detailed best result comparison
    print("\n" + "="*100)
    print("TEST 3: BEST RESULT COMPARISON")
    print("="*100 + "\n")

    query = 'Гражданская абарона'
    print(f"Query: '{query}' (misspelled)\n")

    for nocorrect_val in [False, True]:
        search_result = client.search(query, nocorrect=nocorrect_val, type_='all')

        print(f"\nnocorrect={nocorrect_val}:")
        print(f"  Misspell corrected: {search_result.misspell_corrected}")
        print(f"  Original: {search_result.misspell_original}")
        print(f"  Result:   {search_result.misspell_result}")

        if search_result.best:
            best = search_result.best
            print(f"  Best result type: {best.type}")
            if hasattr(best.result, 'name'):
                print(f"  Best result: {best.result.name}")
            elif hasattr(best.result, 'title'):
                print(f"  Best result: {best.result.title}")

    print("\n" + "="*100)
    print("✓ Tests completed successfully!")
    print("="*100 + "\n")

if __name__ == '__main__':
    main()

File: test/test_search_nocorrect.pl
#!/usr/bin/perl
=head1 NAME

test_search_nocorrect.pl - Test the impact of nocorrect parameter on search results

=head1 DESCRIPTION

This script tests how the 'nocorrect' parameter affects search results in Yandex Music API:

1. Compare search results with nocorrect=true vs nocorrect=false
2. Shows misspelling detection and correction
3. Tests search_suggest method
4. Shows impact of quotes on search results
5. Analyzes best result differences

=head1 USAGE

    perl test/test_search_nocorrect.pl

Requirements:
    - test/token.txt with valid Yandex Music OAuth token
    - perl modules: JSON::XS, LWP::UserAgent, URI::URL, Data::Dumper

=head1 AUTHOR

Test script for Yandex Music LMS Plugin

=cut

use strict;
use warnings;
use JSON::XS;
use LWP::UserAgent;
use HTTP::Request;
use File::Basename 'dirname';
use File::Spec;
use URI::URL qw(url);
use Data::Dumper;

# Read token from test/token.txt
my $script_dir = dirname(__FILE__);
my $token_file = File::Spec->catfile($script_dir, 'token.txt');

if (! -e $token_file) {
    die "Token file not found at: $token_file\n";
}

open my $fh, '<', $token_file or die "Cannot read token file: $!\n";
my $token = do { local $/; <$fh> };
close $fh;
chomp($token);

die "Token file is empty\n" if !$token || $token =~ /^\s*$/;

print "Token: " . substr($token, 0, 10) . "...\n\n";

# Initialize LWP UserAgent
my $ua = LWP::UserAgent->new();
$ua->timeout(15);

my $base_url = 'https://api.music.yandex.net';

# Test queries
my @test_queries = (
    {
        name => 'Correct spelling',
        query => 'Гражданская оборона',
    },
    {
        name => 'Misspelled (missing letter)',
        query => 'Гражданская абарона',
    },
    {
        name => 'Misspelled (extra letter)',
        query => 'Би-2 Сонг',
    },
);

sub make_search_request {
    my ($query, $nocorrect, $type) = @_;
    $type //= 'all';

    my $params = {
        'text'             => $query,
        'nocorrect'        => $nocorrect ? 'True' : 'False',
        'type'             => $type,
        'page'             => 0,
        'playlist-in-best' => 'True',
    };

    my $url = url("$base_url/search");
    $url->query_form(%$params);

    my $req = HTTP::Request->new(GET => $url->as_string);
    $req->header('Authorization' => "OAuth $token");
    $req->header('User-Agent' => 'Yandex-Music-API');
    $req->header('Accept-Language' => 'ru');

    print "  [nocorrect=" . ($nocorrect ? 'true' : 'false') . "] Requesting: " . $url->as_string . "\n";

    my $res = $ua->request($req);

    if (!$res->is_success) {
        print "    ERROR: " . $res->status_line . "\n";
        return undef;
    }

    my $data = eval { decode_json($res->content) };
    if ($@) {
        print "    ERROR parsing JSON: $@\n";
        return undef;
    }

    return $data->{result};
}

sub print_separator {
    my ($char, $length) = @_;
    $char //= '=';
    $length //= 100;
    print $char x $length . "\n";
}

sub format_counts {
    my ($result) = @_;

    if (!$result) {
        return {};
    }

    return {
        tracks           => $result->{tracks} ? $result->{tracks}->{total} : 0,
        artists          => $result->{artists} ? $result->{artists}->{total} : 0,
        albums           => $result->{albums} ? $result->{albums}->{total} : 0,
        playlists        => $result->{playlists} ? $result->{playlists}->{total} : 0,
        videos           => $result->{videos} ? $result->{videos}->{total} : 0,
        podcasts         => $result->{podcasts} ? $result->{podcasts}->{total} : 0,
        podcast_episodes => $result->{podcast_episodes} ? $result->{podcast_episodes}->{total} : 0,
        users            => $result->{users} ? $result->{users}->{total} : 0,
    };
}

sub compare_search_results {
    my ($query) = @_;

    print "\n" . "="x100 . "\n";
    print "Query: '$query'\n";
    print_separator();

    my %results;
    for my $nocorrect (0, 1) {
        my $result = make_search_request($query, $nocorrect, 'all');
        $results{$nocorrect} = $result;
    }

    print_separator('─');

    # Print comparison table
    printf "%-20s %-30s %-30s %-30s\n", 'Category', 'nocorrect=false', 'nocorrect=true', 'Δ (difference)';
    print_separator('─');

    my @types = qw(tracks artists albums playlists videos podcasts podcast_episodes users);

    for my $type (@types) {
        my $count_false = $results{0} ? $results{0}->{$type}->{total} // 0 : 0;
        my $count_true = $results{1} ? $results{1}->{$type}->{total} // 0 : 0;

        if ($count_false > 0 || $count_true > 0) {
            my $diff = $count_true - $count_false;
            my $diff_str = $diff > 0 ? "+$diff" : $diff == 0 ? "0" : "$diff";
            printf "%-20s %-30s %-30s %-30s\n", $type, $count_false, $count_true, $diff_str;
        }
    }

    print_separator('─');

    # Print misspelling info
    print "\nMISPELLING INFORMATION:\n";
    print_separator('─');

    for my $nocorrect (0, 1) {
        my $res = $results{$nocorrect};
        if (!$res) {
            print "  nocorrect=" . ($nocorrect ? 'true' : 'false') . ": No results\n";
            next;
        }

        printf "nocorrect=%s:\n", $nocorrect ? 'true' : 'false';
        printf "  Original query:     %s\n", $res->{misspell_original} // 'N/A';
        printf "  Corrected to:       %s\n", $res->{misspell_result} // 'N/A';
        printf "  Was corrected:      %s\n", $res->{misspell_corrected} ? 'yes' : 'no';
        printf "  nocorrect param:    %s\n", $res->{nocorrect} // 'N/A';

        # Print best result
        if ($res->{best}) {
            printf "  Best result type:   %s\n", $res->{best}->{type};
            if ($res->{best}->{result}) {
                my $result = $res->{best}->{result};
                if ($result->{name}) {
                    printf "  Best result:        %s\n", $result->{name};
                } elsif ($result->{title}) {
                    printf "  Best result:        %s\n", $result->{title};
                }
            }
        }
        print "\n";
    }
}

sub test_search_suggest {
    my ($part) = @_;

    print "\n" . "="x100 . "\n";
    print "SEARCH SUGGEST TEST: '$part'\n";
    print_separator();

    my $url = url("$base_url/search/suggest");
    $url->query_form('part' => $part);

    my $req = HTTP::Request->new(GET => $url->as_string);
    $req->header('Authorization' => "OAuth $token");
    $req->header('User-Agent' => 'Yandex-Music-API');

    print "Requesting: " . $url->as_string . "\n";

    my $res = $ua->request($req);

    if (!$res->is_success) {
        print "ERROR: " . $res->status_line . "\n";
        return;
    }

    my $data = eval { decode_json($res->content) };
    if ($@) {
        print "ERROR parsing JSON: $@\n";
        return;
    }

    my $result = $data->{result};
    if (!$result || !$result->{suggests}) {
        print "No suggestions found\n";
        return;
    }

    print_separator('─');
    printf "Found %d suggestions:\n\n", scalar(@{$result->{suggests}});

    my $count = 0;
    for my $suggest (@{$result->{suggests}}) {
        $count++;
        last if $count > 10;
        printf "  %2d. %s\n", $count, $suggest;
    }
    print "\n";
}

# ===== MAIN TEST EXECUTION =====

print "="x100 . "\n";
print "YANDEX MUSIC SEARCH NOCORRECT PARAMETER TEST\n";
print "="x100 . "\n";

print "\nTEST 1: NOCORRECT PARAMETER IMPACT\n";
for my $test_query (@test_queries) {
    compare_search_results($test_query->{query});
}

print "\n" . "="x100 . "\n";
print "TEST 2: SEARCH_SUGGEST METHOD\n";
print "="x100 . "\n";

test_search_suggest('граж');
test_search_suggest('би');

print "\n" . "="x100 . "\n";
print "✓ Tests completed successfully!\n";
print "="x100 . "\n\n";
